### ISSUE TITLE
"Could not add/remove hint" error using nndb migration on Windows

### ISSUE THREAD

**Issue by DanTup** [dart-lang/sdk#43178]
Using Dart SDK at N:\Apps\Dart\nightly-2020-08-25
Dart SDK version: 2.10.0-edge.3e15656b3ef89516a36e0b3d66150e881ba2fa8d (be) (Tue Aug 25 12:15:29 2020 +0000) on "windows_x64"

- git clone https://github.com/DanTup/dart-repro-migrate
- cd dart-repro-migrate
- N:\Apps\Dart\nightly-2020-08-25\bin\pub get
- N:\Apps\Dart\nightly-2020-08-25\bin\dart migrate --skip-pub-outdated
- Click lib/my_migrate_project.dart in the project list
- Click "line 10: Added a non-null assertion to nullable expression" (the single item in the last list)
- ## Issue 1: Bad path shown:
  > Map​.entries (..\\..\\..\\..\\..\\N:\Apps\Dart\nightly-2020-08-25\lib\core\map.dart:241): type argument 0 of return type of Map​.entries (map​.dart:241:32)
- Click the "..\\..\\..\\..\\..\N:\Apps\Dart\nightly-2020-08-25\lib\core\map.dart:241" link
- Click "Add `/*!*/` hint" button
- ## Issue 2: Error notification pops up:
  (Note it's looking in the SDK folder for   `my_migrate_project.dart`)
  > Could not add/remove hint
  > FileSystemException(path=N:\Apps\Dart\nightly-2020-08-25\lib\core\my_migrate_project.dart; message=Cannot open file)
  

![Untitled](https://user-images.githubusercontent.com/1078012/91178875-b1995b80-e6dd-11ea-8a8f-e89f7d7678af.png)

@srawlins

**Comment by devoncarew** [comment]
@srawlins - can you take a look? I don't know if this version (for the bug report) contains your recent fixes.

**Comment by DanTup** [comment]
I believe it should include b1094db and e7c79ae (it was the nightly built from 3e15656b3ef89516a36e0b3d66150e881ba2fa8d around 8 hours ago).

**Comment by srawlins** [comment]
Today is a busy day, but I plan on making "Migrator on Windows" my priority for the rest of the week.

**Comment by srawlins** [comment]
This is forked from Danny's comments here: https://github.com/dart-lang/sdk/issues/42636, which may have more context.

**Comment by srawlins** [comment]
@DanTup Thanks for the excellent repro steps. I'm having trouble reproducing in unit tests (on a Mac :/) but I'm wondering if the drive letter is related. Your Dart SDK (and so `Map.entries`) are on `N:`; what about the cloned package, and your current working directory? Also `N:`? A different drive?

**Comment by DanTup** [comment]
> Your Dart SDK (and so Map.entries) are on `N:`; what about the cloned package, and your current working directory? Also `N:`? A different drive?

Ah sorry, I didn't include the path to that. It's on a different drive at `M:\Dev\TestStuff\dart_migrate_issue`

**Comment by srawlins** [comment]
Oh thank goodness. I can reproduce. 🎉 🥳 🎉 

It does have to do with separate drive letters.

(Note: in unit tests, I am finding more odd things; an assertion fails, or the test wants to do several casts and turn `() => null` into `() => null!`; I think these are all unrelated, but you still might run into them while migrating.)

**Comment by DanTup** [comment]
Great, thanks! I was able to get things running from source (so don't need to wait for a nightly build) so if you let me know when you think you have a fix, I can do some more testing.

**Comment by DanTup** [comment]
@srawlins btw, slightly off-topic.. would it be reasonable to add a `--machine` mode that puts command line output into JSON (eg. for the URL to launch, and also an error if the pub check fails)? That way we could add a command to the IDEs that launches this automatically without dropping to a terminal. If the pub check fails, we could show a notification with an option to review the output, or "run anyway" (which just re-runs with the --skip flag).

We could do it by parsing text, but JSON would be much better.

**Comment by srawlins** [comment]
@DanTup I've requested this at https://github.com/dart-lang/sdk/issues/43224

**Comment by srawlins** [comment]
I think I have fixed the bug, but the Windows test bots are giving different results than my local tests, so I'm re-opening to address those tests.

**Comment by DanTup** [comment]
Let me know if/when you want me to do any testing on my PC here!

**Comment by DanTup** [comment]
@srawlins I still see this on the latest nightly:

Dart SDK version: 2.10.0-edge.42c4c3c1bdc9d103ec6cbbc3ddc3047538a3aa37 (be) (Mon Sep 7 12:36:13 2020 +0000) on "windows_x64"

The path looks less bad, but still has a leading slash:

![image](https://user-images.githubusercontent.com/1078012/92398991-6fcad500-f121-11ea-9982-f480c9ee1904.png)

But following the instructions above still results in the same error (it's still looking inside the SDK for my files):

![image](https://user-images.githubusercontent.com/1078012/92399029-853fff00-f121-11ea-9b1a-c8e663158ba1.png)

**Comment by srawlins** [comment]
I'm working on a good Windows situation to tackle this again.

**Comment by DanTup** [comment]
👍 

Let me know if you want me to test anything here. I'm not set up to build the SDK, but can use the nightlies and can run the migrate tool from latest source (with previous nights VM build).

**Comment by srawlins** [comment]
I've got a Windows setup now and the steps all work for me:

* `dart.exe --version`: Dart SDK version: 2.10.0-edge.e89279c85f80d57889a8a15757699c951200b391 (be) (Tue Sep 22 21:59:31 2020 +0000) on "windows_x64"
* Sample project is at `E:\dart-repro-migrate`
* `C:\Users\srawlins\dart-sdk\sdk\out\ReleaseX64\dart-sdk\bin\dart migrate --skip-pub-outdated`

1. Click lib/my_migrate_project.dart in the project list

   All good

2. Click "line 10: Added a non-null assertion to nullable expression" (the single item in the last list)

   The URL shown in the status bar is: 127.0.0.1:59486/E:/dart-repro-migrate/lib/my_migrate_project.dart?line=10&offset=409&authToken=xyz. It links fine.

3. Click the "..\..\..\..\..\N:\Apps\Dart\nightly-2020-08-25\lib\core\map.dart:241" link

   The link now says:

   > Map​.entries (/C:/Users/srawlins/dart-sdk/sdk/out/ReleaseX64/dart-sdk/lib/core/map.dart:241): type argument 0 of return type of Map​.entries (map​.dart:241:32)

   (The slash before `C:` is fine.)

4. Click "`Add /*!*/ hint`" button

   The URL shown in the status bar is: 127.0.0.1:59609/E:/dart-repro-migrate/lib/my_migrate_project.dart?offset=355&end=355&replacement=%2F%2A%21%2A%2F. It links fine.

**Comment by franklinyow** [comment]
Any work left before we can close this?

**Comment by srawlins** [comment]
We should see if it is still reproducible.

@DanTup can you still reproduce this on your end, with a recent version of Dart/Flutter?

**Comment by DanTup** [comment]
Sorry, I missed some of the previous notifications. I tested this again, and do not see the error anymore. Some minor issues I did see though (lmk if these are better in their own issues - or are expected/known):

1. Inconsistent path formats

![path_formats](https://user-images.githubusercontent.com/1078012/96134218-1eb5ba00-0ef3-11eb-9451-dc09aa0bebc6.png)

2. Weird text rendering issues

There's some weird ghosting of text - looks like it's being rendered twice in different colours with the positions drifting out of sync.

![text_ghosting](https://user-images.githubusercontent.com/1078012/96134527-31c88a00-0ef3-11eb-8784-f9271169f03a.png)

3. Changes made on disk before clicking "Apply Migration"

When I clicked "Add /*!*/" the file was immediately modified on disk. Based on the dialog that appears when clicking "Apply Migration", I was expecting it to only make changes when I clicked that button. I'm not sure if this is intended?
